### ISSUE TITLE
Custom notification sound for iOS

### ISSUE THREAD

**Issue by Diggory** [Ranchero-Software/NetNewsWire#3704]
I always appreciate it when an app uses a custom notification sound, so you can tell which app has sent a notification without having to look at actual devices. Especially since Apple's News app uses the default tone.

 NNW currently uses the default tone, it would be useful to use a custom tone so that the user can differentiate whether 'news' is coming from NNW or somewhere else.

I suggest using a CC licensed sound, unless the NNW community can generate our own.  (I am not a musician/tone composer!)

**Comment by Diggory** [comment]
I asked in the #nnw-ios-feedback for opinions about this, but got no reply, so I will implement this and create a PR.

**Comment by vincode-io** [comment]
Never mind, I opened up slack and you are way ahead of me.

**Comment by Diggory** [comment]
Yes, sorry about that, I should have read the doc first...

**Comment by vincode-io** [comment]
Why would we only implement a custom sound for iOS only?

**Comment by Diggory** [comment]
I'd forgotten about macOS when adding the feature, but then I noticed that the notification code is shared by both platforms.  However, I also saw that on macOS the app doesn't make any notification sound at the moment and doesn't have the option to in System Prefs.

I thought that the sound was more important on iOS when often you aren't looking at the device, but I'm happy to add the sound to macOS as well if you'd like.  Notification sounds are rather less frequent on macOS though.  

Happy either way and to be guided by you.  Some people may find the noise annoying on macOS however.

**Comment by vincode-io** [comment]
I wonder why Apple doesn't make the sound customizable on a per app basis in the Notification settings to begin with?

*If* we add something like this, it may make sense to let the user select the sound they want to use. Things like this are very subjective. That would include the option to not have a notification sound at all if we can't already do so in the system Settings on macOS.

**Comment by Diggory** [comment]
The user can disable notification sounds per-app in the System prefs notification pane.

Good idea about multiple choice on the sound.

**Comment by Diggory** [comment]
> I wonder why Apple doesn't make the sound customizable on a per app basis in the Notification settings to begin with?

I suppose because the sounds can be changed programmatically.  So there's not one specific notification sound per app.  It's not a 1-to-1 relationship.

**Comment by vincode-io** [comment]
The Mac app does have the potential to do sounds and offer it as an option in settings. You just have to pass the right options in `UNUserNotificationCenter.current().requestAuthorization`. You also have to add the sound file to the Mac target.

<img width="780" alt="Screen Shot 2022-10-01 at 3 30 24 PM" src="https://user-images.githubusercontent.com/16448027/193427276-2454aa57-7153-4541-af30-9dc4d11af7a7.png">

I wasn't able to get it to work though. I'm not sure if there is a problem on my machine or an issue with UNNotificationSoundName. It never seemed to be able to find the sound file in the App Bundle on macOS.

**Comment by Diggory** [comment]
I had the same experience as you, yesterday. I added the sound option in the requestAuthorisation function and added the sound to the macOS target but for the life of me couldn’t get anything other than the tritone as the sound. 

I made a simple example project which only sends notifications but had the same issue. I’ll ask on the Apple Dev forums and hopefully someone will point out that there’s some hoop you have to jump through that’s undocumented. 😔

**Comment by vincode-io** [comment]
I hope we get an answer. I was so frustrated and pissed off that I was going to decompile `UNNotificationSoundName` to see where the heck it was looking inside the bundle. I couldn't even get it to pick up the file in any of the `/Library/Sounds` locations that it is documented to look in.

**Comment by Diggory** [comment]
Made a minimal test case and posted to the dev forums...  Fingers crossed.

https://developer.apple.com/forums/thread/716650

**Comment by Diggory** [comment]
I've updated my test app (https://github.com/Diggory/macOSLocalNotifications) with multiple sounds in differing formats.  None of them appears to work.  Except for one MP3.  I can't work out what is so special about it.

**Comment by Diggory** [comment]
I was trying to think of custom alert sounds in other apps, but I could only think of Slack.  They appear to use MP3s for their sounds.

**Comment by vincode-io** [comment]
I was able to get your app to work with everything except the .caf files. I did have to enable "Play user interface sound effects" to get it to work. It looks like it is supposed to be required now. Check the SlackHQ reply to this: https://twitter.com/bansalvks/status/1481305026257879040?s=21&t=MtV4zLN0wgodUYhbpj7Xtg

I was finally able to get NetNewsWire to play `notificationSoundBlip.mp3`.

I had to delete the NetNewsWire.app and reboot first though. I also removed its files under `~/Library/Containers`, but I don't think that had anything to do with it. I restored everything there and it will still play `notificationSoundBlip.mp3`.

I suspect that if you enabled Notifications before Monterey and migrated that there is some legacy setting in macOS that keeps you getting the tritone even if you have  "Play user interface sound effects". This is probably something Apple did to try to make the transition to the new settings in Monterey easier on users.
